File: Structured Leetcode/4 Linked List/reverseLinkedListsII.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseBetween(head, left, right):
        newList = None
        finalNode =  head
        cur = head

        while True:
            if cur.val == left:
                finalNode = head.next 
                newList = cur
            
            else:
                copy = cur
                copy.next = newList
                newList = copy                

                if newList.val == right:
                    finalNode.next = newList
                    newList.next = cur.next
                    break
            cur = cur.next

        return head


a = ListNode(5)
b = ListNode(4, a)
c = ListNode(3, b)
d = ListNode(2, c)
head = ListNode(1, d)

print(Solution.reverseBetween(head, 2, 4).val)

File: Structured Leetcode/4 Linked List/reverseLinkedListsII.c
#include <stdbool.h>


// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* reverseBetween(struct ListNode* head, int left, int right) {
    struct ListNode newHead;

    newHead.next = head;
    struct ListNode *cur = head;
    struct ListNode *prev = &newHead;
    struct ListNode *interlist;
    struct ListNode *interListTail;


    while(true) {

        if (cur->val == left) {
            interlist = cur;
            interListTail = cur;

        } else if (cur->val == right) {
            struct ListNode copy;
            copy.val = cur->val;
            copy.next = interlist;
            interlist = &copy;

            prev->next = interlist;
            interListTail->next = cur->next;
            break;

        } else if (interlist != 0) {
            struct ListNode copy;
            copy.val = cur->val;
            copy.next = interlist;
            interlist = &copy;
        }
    
        cur = cur->next;
        if (interlist == 0) {
            prev = prev->next;
        }
        
    }
    return newHead.next;
};

int main() {
    struct ListNode a;
    a.val = 5;
    a.next = 0;

    struct ListNode b;
    b.val = 4;
    b.next = &a;

    struct ListNode c;
    c.val = 3;
    c.next = &b;

    struct ListNode d;
    d.val = 2;
    d.next = &c;

    struct ListNode head;
    head.val = 1;
    head.next = &d;

    // struct ListNode* head2 = reverseBetween(&head, 2, 4);
    struct ListNode* head2 = reverseBetween(&head, 1, 3);

    return 0;
};